crates/wasmparser/src/readers/core/types/matches.rs

```diff
@@ -103,27 +103,28 @@ impl<'a> Matches for WithRecGroup<&'a SubType> {
 
 impl<'a> Matches for WithRecGroup<&'a CompositeType> {
     fn matches(types: &TypeList, a: Self, b: Self) -> bool {
-        match (&*a, &*b) {
-            (CompositeType::Func(fa), CompositeType::Func(fb)) => Matches::matches(
+        use CompositeInnerType::*;
+        match (&(*a).inner, &(*b).inner) {
```

Comment:
Shouldn't matching take `shared`ness into account? Does an unshared `T` match a shared `T`? I vaguely remember reading some email about starting conservatively and saying no, but I don't remember...

Reply:
Yeah, that's a good point... let me think about that.

Reply:
Seems fine to start by only allowing shared things to match other shared things and similarly for unshared. We can land code that does that in `wasm-tools` for now, in the name of incrementality, even if the upstream proposal wants to do something more complicated.

Reply:
Ok, I did the simple thing for now and opened: https://github.com/WebAssembly/shared-everything-threads/issues/72.

Reply:
Eventually I would like to upstream these related spec tests: 
 - https://github.com/WebAssembly/binaryen/blob/main/test/spec/shared-struct.wast
 - https://github.com/WebAssembly/binaryen/blob/main/test/spec/shared-array.wast
 - https://github.com/WebAssembly/binaryen/blob/main/test/spec/shared-polymorphism.wast

Reply:
Nice! I tried to drop those in and I think I'm seeing parsing differences between Binaryen and wasm-tools?!

```
---- "tests/local/shared-everything-threads/array.wast" ----
failed test: tests/local/shared-everything-threads/array.wast

Caused by:
    0: failed to run `json-from-wast` cli subcommand
    1: failed to run "/home/abrown/Code/wasm-tools/target/debug/wasm-tools" "json-from-wast" "tests/local/shared-everything-threads/array.wast" "--pretty" "--wasm-dir" "/tmp/.tmpjHLO2v"
       stdout:
       stderr: error: unexpected token, expected one of: `funcref`, `externref`, `exnref`, `anyref`, `eqref`, `structref`, `arrayref`, `i31ref`, `nullfuncref`, `nullexternref`, `nullexnref`, `nullref`, left paren
            --> tests/local/shared-everything-threads/array.wast:78:8
             |
          78 |   (elem)
             |        ^


---- "tests/local/shared-everything-threads/struct.wast" ----
failed test: tests/local/shared-everything-threads/struct.wast

Caused by:
    0: failed to run `json-from-wast` cli subcommand
    1: failed to run "/home/abrown/Code/wasm-tools/target/debug/wasm-tools" "json-from-wast" "tests/local/shared-everything-threads/struct.wast" "--pretty" "--wasm-dir" "/tmp/.tmpN86fk1"
       stdout:
       stderr: error: expected `(`
            --> tests/local/shared-everything-threads/struct.wast:56:27
             |
          56 |     (type (shared (struct anyref)))
             |                           ^
```

@fitzgen, I'm not the GC expert here... do we need to improve parsing of GC here as well or should we touch these tests up to make them parse?

Reply:
I added the tests but modified them just enough to avoid the missing `wast` parsing (see commit message of ef101c3).